rfc/system/5606-json-idl.md

```diff
@@ -0,0 +1,343 @@
+
+#### Dataclass Transformer
+##### Before
+###### Convert Python Value to Literal
+Uses `mashumaro JSON Encoder` to turn a dataclass value to a JSON string, and store it to protobuf `Struct` .
+Note: For `FlyteTypes`, we will inherit mashumaro `SerializableType` to define our own serialization behavior, which includes upload file to remote storage.
+
+###### Convert Literal to Python Value
+Uses `mashumaro JSON Decoder` to turn a JSON string to a python value, and recursively fixed int attributes to int (it will be float because we stored it in to `Struct`).
+
+Note: For `FlyteTypes`, we will inherit the mashumaro `SerializableType` to define our own serialization behavior, which includes uploading files to remote storage.
+##### After
+###### Convert Python Value to Literal
+Uses `msgpack.dumps()` to turn a Python value into a byte string.
+Note: For `FlyteTypes`, we will need to customize serialization behavior by msgpack reference here.
+
+https://github.com/msgpack/msgpack-python?tab=readme-ov-file#packingunpacking-of-custom-data-type 
+
+###### Convert Literal to Python Value
+
+Uses `msgpack.loads()` to turn a byte string into a Python value.
+
+Note: For `FlyteTypes`, we will need to customize deserialization behavior by `msgpack` reference here.
+
+https://github.com/msgpack/msgpack-python?tab=readme-ov-file#packingunpacking-of-custom-data-type 
+
+
+#### Pydantic Transformer
+##### Before
+###### Convert Python Value to Literal
+Convert `BaseModel` to a JSON string, and then convert it to a Protobuf `Struct`.
+###### Convert Literal to Python Value
+Convert Protobuf `Struct` to a JSON string and then convert it to a `BaseModel`.
+##### After
+###### Convert Python Value to Literal
+Convert the Pydantic `BaseModel` to a JSON string, then convert the JSON string to a `dictionary`, and finally, convert it to a `byte string` using msgpack.
+###### Convert Literal to Python Value
+Convert `byte string` to a `dictionary` using `msgpack`, then convert dictionary to a JSON string, and finally, convert it to Pydantic `BaseModel`.
+
+### FlyteCtl
+In Flytectl, we can construct input for the execution, so we have to make sure the values we passed to FlyteAdmin can all be constructed to Literal.
+
+https://github.com/flyteorg/flytectl/blob/131d6a20c7db601ca9156b8d43d243bc88669829/cmd/create/serialization_utils.go#L48 
+
+### FlyteConsole
+#### Show input/output on FlyteConsole
+We will get node’s input output literal value by FlyteAdmin’s API, and get the json byte string in the literal value.
+
+We can use MsgPack dumps the json byte string to a dictionary, and shows it to the flyteconsole.
+#### Construct Input
+We should use `msgpack.encode` to encode input value and store it to the literal’s json field.
+
+
+
+## 4 Metrics & Dashboards
+
+None
+
+## 5 Drawbacks
+There's no breaking changes if we use `Annotated[dict, Json]`, but we need to be very careful about will there be any breaking changes.
+
+## 6 Alternatives
```

Comment:
Let me clarify the serialization process (deserialization is the reverse):

- Using msgpack:
basemodel -> json string -> dictionary -> msgpack bytes

- Using utf-8:
basemodel -> json string -> json byte string

Reply:
But the benefit is that if we use `msgpack`, we can have smaller data bytes when uploading to remote storage and downloading from it.
This is really a HUGE benefit and lots of people wants it.

Reply:
UTF-8 and msgpack are not the only possible formats, right? Can we add more color to this? What if of instead dictating msgpack as the serialization format everywhere we added that as an explicit field as part of the new literal type? This way we could have more specialized serialization formats used by plugins in case they want/need to.